opensearch-operator/pkg/reconcilers/tls.go

```diff
@@ -102,50 +102,98 @@ func (r *TLSReconciler) handleTransport() error {
-		}
+		ca, err := r.getCACert()
 		if err != nil {
 			return err
 		}
-
-		adminSecret := corev1.Secret{}
-		if err := r.Get(r.ctx, client.ObjectKey{Name: adminSecretName, Namespace: namespace}, &adminSecret); err != nil {
-			adminCert, err := ca.CreateAndSignCertificate("admin", clusterName, nil)
-			if err != nil {
-				r.logger.Error(err, "Failed to create admin certificate", "interface", "transport")
-				r.recorder.AnnotatedEventf(r.instance, map[string]string{"cluster-name": r.instance.GetName()}, "Warning", "Security", "Failed to create admin certificate")
-				return err
-			}
-			adminSecret = corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: adminSecretName, Namespace: namespace}, Type: corev1.SecretTypeTLS, Data: adminCert.SecretData(ca)}
-			if err := ctrl.SetControllerReference(r.instance, &adminSecret, r.Client.Scheme()); err != nil {
-				return err
-			}
-			if err := r.Create(r.ctx, &adminSecret); err != nil {
-				r.logger.Error(err, "Failed to store admin certificate in secret", "interface", "transport")
-				return err
-			}
+		err = r.createAdminSecret(ca)
+		if err != nil {
+			return err
 		}
-		// Add admin_dn to config
 		r.reconcilerContext.AddConfig("plugins.security.authcz.admin_dn", fmt.Sprintf("[\"CN=admin,OU=%s\"]", clusterName))
-	} else {
-		// Add provided admin_dn to config
-		adminDn := strings.Join(tlsConfig.AdminDn, "\",\"")
-		r.reconcilerContext.AddConfig("plugins.security.authcz.admin_dn", fmt.Sprintf("[\"%s\"]", adminDn))
+		return nil
 	}
+
+	adminDn := strings.Join(tlsConfig.AdminDn, "\",\"")
+	r.reconcilerContext.AddConfig("plugins.security.authcz.admin_dn", fmt.Sprintf("[\"%s\"]", adminDn))
 	return nil
 }
 
+func (r *TLSReconciler) securityChangeVersion() bool {
+	newVersionConstraint, err := semver.NewConstraint(">=2.0.0")
+	if err != nil {
+		panic(err)
+	}
+
+	version, err := semver.NewVersion(r.instance.Spec.General.Version)
+	if err != nil {
+		r.logger.Error(err, "unable to parse version, assuming >= 2.0.0")
+		return true
+	}
+	return newVersionConstraint.Check(version)
+}
+
+func (r *TLSReconciler) adminCAProvided() bool {
+	if r.securityChangeVersion() {
+		return r.instance.Spec.Security.Tls.Http.TlsCertificateConfig.CaSecret.Name != ""
+	}
+	return r.instance.Spec.Security.Tls.Transport.TlsCertificateConfig.CaSecret.Name != ""
+}
+
+func (r *TLSReconciler) providedCAForAdminCert() (tls.Cert, error) {
+	if r.securityChangeVersion() {
+		return r.providedCaCert(
+			r.instance.Spec.Security.Tls.Http.TlsCertificateConfig.CaSecret.Name,
+			r.instance.Namespace,
+		)
+	}
+	return r.providedCaCert(
+		r.instance.Spec.Security.Tls.Transport.TlsCertificateConfig.CaSecret.Name,
+		r.instance.Namespace,
+	)
+}
+
+func (r *TLSReconciler) getCACert() (tls.Cert, error) {
+	if r.adminCAProvided() {
+		return r.providedCAForAdminCert()
+	}
+	return util.ReadOrGenerateCaCert(r.pki, r.Client, r.ctx, r.instance)
+}
+
+func (r *TLSReconciler) createAdminSecret(ca tls.Cert) error {
```

Comment:
This method should check if the admin secret already exists before creating a new one. Otherwise it will generate new ones on each reconciliation run, which is quite cpu-intensive.

Reply:
Need to have a bit of a think about this.  The issue here is that if you're using intermediate CAs, and they get regenerated due to expiry you also need to regenerate the admin certificate.